FutureNorthants/northants-website
# Implement Incremental Static Regeneration

This PR implements [ISR in Next.js](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration). 

The main benefit of this is greatly improved performance serving pages as it returns pre-rendered HTML for most requests without the need to request data from the CMS.

The first time a page is visited, a static page will be generated. Each subsequent request will serve the pre-generated page. After the revalidate time (in seconds) has passed it will do a new query to the server to get the latest data and update the HTML. 

The revalidate time is currently set to 10 seconds so you can test it out. In production it may be worth setting to a larger value. 

Running `yarn dev` will fetch the data each time the page is refreshed so you can see your changes. 
Running `yarn build && yarn start` will use ISR. 

## To Test
- Checkout this branch, then `yarn install`
- Run end to end tests `yarn e2e:headless` and check for any errors
- Run component tests `npx cypress run-ct` and check for any errors
- `yarn build && yarn start`
- Open the dev tools and go to network tab. Visit a website page and notice the response time. 
- Refresh the page and notice the shorter response time. The status should also be 304. 
- Wait over 10 seconds, refresh the page and it should return a 200 status and the latest version of the page

You can also try increasing the revalidate time, visiting a page, then editing the page in the CMS, then refresh the frontend to see the cached version. Wait for the revalidate time then refresh and the updated version should be displayed.
